toolchains/toolchain.bzl

```diff
@@ -1,78 +1,144 @@
+
+LINUX_AMD64 = ["@platforms//os:linux", "@platforms//cpu:x86_64"]
+LINUX_ARM64 = ["@platforms//os:linux", "@platforms//cpu:aarch64"]
+MAC_AMD64 = ["@platforms//os:osx", "@platforms//cpu:x86_64"]
+MAC_ARM64 = ["@platforms//os:osx", "@platforms//cpu:aarch64"]
+WINDOWS_AMD64 = ["@platforms//os:windows", "@platforms//cpu:x86_64"]
+WINDOWS_ARM64 = ["@platforms//os:windows", "@platforms//cpu:aarch64"]
+
+NO_REMOTE_EXECUTION = {None: []}
+DEFAULT_EXEC_CONSTRAINT_GROUPS = {
+    "linux_amd64": LINUX_AMD64,
+    "linux_arm64": LINUX_ARM64,
+    "mac_amd64": MAC_AMD64,
+    "mac_arm64": MAC_ARM64,
+    "windows_amd64": WINDOWS_AMD64,
+    "windows_arm64": WINDOWS_ARM64,
+}
+
+_NO_EXEC_CONSTRAINT_GROUPS_ERROR = """You did not specify a set of execution platforms for your toolchain.
+
+This is needed if you are using remote execution. If you don't intend to use remote execution, set exec_constraint_groups = NO_REMOTE_EXECUTION
+
+If you do have remote execution, each execution platform (both remote and local) that might result in a different configuration (eg. different tool binaries) should be listed.
+
+A reasonable set of defaults is DEFAULT_EXEC_PLATFORMS, which provides amd64 and arm64 variants for linux, mac, and windows.
+
+For example, if all your devs use linux amd64 machines locally, and your remote execution machines are linux arm64 machines, you would write:
+```
+{
+    "linux_amd64": ["@platforms//os:linux", "@platforms//cpu:x86_64"]
+    "linux_arm64": ["@platforms//os:linux", "@platforms//cpu:aarch64"]
+}
+```
+"""
+
+def _create_toolchain(*, label, enabled_features, tool_map):
+    capabilities_to_features = {}
+    defaults = {}
+    for action_type, tool in tool_map.configs.items():
+        capabilities = tuple(sorted([cap.label for cap in tool.capabilities]))
+        if capabilities not in capabilities_to_features:
+            capabilities_to_features[capabilities] = get_features(list(enabled_features) + [cap.feature for cap in tool.capabilities], [], [])
+        features = capabilities_to_features[capabilities]
+        defaults[action_type] = resolve_action(action_type = action_type, features = features, extra_args = [])
+
+    return ToolchainConfigInfo(
+        label = label,
+        enabled_features = enabled_features,
+        defaults_by_action = defaults,
+        tool_map = tool_map,
     )
 
-    # Export all the providers inside our ToolchainInfo
-    # so the resolved_toolchain rule can grab and re-export them.
-    toolchain_info = platform_common.ToolchainInfo(
-        ToolchainsInfo = toolchains_info,
-        template_variables = template_variables,
-        default = default,
+def _toolchain_config_impl(ctx):
+    if ctx.attr.features:
+        fail("Features is a reserved attribute in bazel. Did you mean 'default_features'?")
+
+    toolchain_config = _create_toolchain(
+        label = ctx.label,
+        enabled_features = collect_features([target[FeatureSetInfo] for target in ctx.attr.default_features]),
```

Comment:
This looks like the only place in which a "set" is frequently accessed by consuming rules. Do we really need `collect_features` and the small set abstraction? I may be missing something, but couldn't we just use `depset`s internally and call `to_list()` in this single location?